[RDY] Refactor patient

Alberth289346 opened the pull request:
Moving hospital administration bits out of the patient. I think there is more, but this seems like a good first step.

Code is mostly copy/paste, but the various self/patient/hospital replacements are easy to mess up.
I'd be fine with postponing after 0.60

TheCycoONE commented:
Just as an update I'm taking your suggestion to postpone this until post 0.60.

TheCycoONE commented on CorsixTH/Lua/hospital.lua:
```diff
@@ -664,7 +664,7 @@ function Hospital:countPatients()
   -- so Town map now takes this information from here.  (If I am wrong, put it back)
   self.patientcount = 0
-  for _, patient in pairs(self.patients) do
+  for _, patient in ipairs(self.patients) do
```
Any reason to switch to ipairs here?

Alberth289346 replied:
Other places use it as a list (which is stupid imho, but I am doing patient rather than hospital atm).

TheCycoONE commented:
924fbaa3c6d3bc82c2b56ab0d542d29b10f67de4 is dubious. A function to remove from a table by value could be useful in utils, repeating the code doesn't seem better to me.

TheCycoONE commented:
3a7f11f61fa4b2b6c85f65a058b5f3f17a6e00c6 is another one I question the value of. Patient may not be the right place for it, but room is worse imho.

TheCycoONE commented:
With the exception for the two commits I've commented I'm very happy with the direction of this patch.

Alberth289346 commented:
just dropped commit 3a7f11f although I disagree with the patient updating hospital administration.

reverted the local function thing of commit 924fbaa too
